sycl/doc/extensions/proposed/sycl_ext_oneapi_free_function_kernels.asciidoc

```diff
@@ -0,0 +1,921 @@
+kernels.
+These implementations automatically insert forward declarations of the free
+function kernels at the top of the translation unit.
+This has ramifications on how the application may declare the free function
+kernels, on the types that may be used in those declarations, and on the way
+the application may reference these kernel identifiers.
+The following example illustrates the forward declarations that the
+implementation inserts:
+
+```
+// Forward declarations of types used by the kernel functions.
+struct mystruct;
+enum myenum : int;
+
+// Each kernel is forward declared in the same namespace in which the
+// application declares it.
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1) void kernel1(int *);
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1) void kernel2(mystruct, myenum);
+
+template<typename T>
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1) void kernel3(T *);
+
+namespace ns {
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1) void kernel4(int *);
+}
+```
+
+As a result, these implementations impose additional restrictions for functions
+that are declared as free function kernels:
+
+* The function must be declared at namespace scope.
+
+* Any type used in the declaration of a parameter must be one of the allowed
+  types listed below.
+
+* If the function is instantiated from a template, any type used to instantiate
+  the template must be one of the allowed types listed below.
+
+* Uses of function identifiers in the application must assume that the free
+  function kernels are forward declared at the top of the translation unit.
+  Note that this can also affect references to functions that are not declared
+  as free functions kernels as illustrated below.
++
+```
+void foo(int) {/*...*/}
+
+void caller() {
+  auto *pf = foo;  // This is ambiguous because foo(float) is forward declared
+                   // in the integration header
+}
+
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void foo(float) {/*...*/}
+```
+
+The allowed types are:
+
+* A {cpp} fundamental type.
+* A class or struct that is defined at namespace scope.
+* A scoped enumeration that is defined at namespace scope.
+* An unscoped enumeration that has an explicit underlying type, where the
+  enumeration is defined at namespace scope.
+* A type alias to one of the above types.
+
+
+== Example
+
+=== Basic invocation
+
+The following example demonstrates how to define a free function kernel and then
+enqueue it on a device.
+
+```
+namespace syclex = sycl::ext::oneapi::experimental;
+
+static constexpr size_t NUM = 1024;
+
+SYCL_EXT_ONEAPI_RANGE_KERNEL(1)
+void iota(float start, float *ptr) {
+  // Get the ID of this kernel iteration.
+  size_t id = syclex::this_kernel::get_id();
+
+  ptr[id] = start + static_cast<float>(id);
+}
+
+void main() {
+  sycl::queue q;
+  sycl::context ctxt = q.get_context();
+
+  // Get a kernel bundle that contains the free function kernel "iota".
+  auto exe_bndl =
+    syclex::get_kernel_bundle<iota, sycl::bundle_state::executable>(ctxt);
```

Comment:
Is there any way to force creation of kernel bundle from SPIR-V or device binary?
When you do i.e. AOT compilation, possibly you have these 2 types of binaries available. Forcing use of SPIR-V can be useful as debug method

Reply:
This is what the second template parameter controls.  In this example, the parameter is `sycl::bundle_state::executable`, which means the bundle will contain native code.  If the ELF file does not contain native code (i.e. it was not AOT compiled), then the SYCL RT will online compile the kernel from SPIR-V.

If the application wants a kernel bundle with SPIR-V, it can do so by using `sycl::bundle_state::input` for the template parameter like this:

```
auto exe_bndl =
    syclex::get_kernel_bundle<iota, sycl::bundle_state::input>(ctxt);
```

Note, however, that there is no guarantee that the ELF file has a SPIR-V representation for every kernel.  Therefore, your application should check to see if the bundle really does contain the kernel.  See this statement from the SYCL spec:

> Since the implementation may not represent all kernels in `bundle_state::input` or `bundle_state::object`, calling this function with one of those states may return a bundle that is missing some of the kernels in `kernelIds`. The application can test for this via `kernel_bundle::has_kernel()`.

(https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_obtaining_a_kernel_bundle)

The current DPC++ behavior when AOT compiling is to *not* store the a SPIR-V copy of the kernel in the ELF file unless you specifically ask for both AOT and SPIR-V.  However, we are thinking of switching the default, so that AOT *does* store a SPIR-V copy by default (unless you specifically ask for it to be omitted).

P.S.: Note that none of this is specific to the extension proposed in this PR.  Everything I said above applies also to regular SYCL kernels.

Reply:
I think this issue is resolved, so I'm marking it "resolved".  @bsochack, if you have more concerns, please feel free to reopen it.